### ISSUE TITLE
Phantomjs crashing when trying to process 100+ links

### ISSUE THREAD

**Issue by saalmaan** [giakki/uncss#250]
Hi Gents, 

I want to scan all the pages in my website to identify unused css, but when I add 100 links to files array phantom js crashes with strange error 

> PhantomJS has crashed. Please read the crash reporting guide at https://github.com/ariya/phantomjs/wiki/Crash-Reporting and file a bug report at https://github.com/ariya/phantomjs/issues/new with the crash dump file attached: /tmp/8ECEAFDC-2355-4665-8293-BF732DEF5F40.dmp

I am using standard implementation

``` js
var uncss = require('uncss');
var files = ['http://example.com/page1.html', 
                  'http://example.com/page2.html', 
                   ...... 
                  '98 more links'],
options = {
 stylesheets: [
   'http://example.com/css/style.css'
 ],
 timeout: 20000,
 report: true
};
uncss(files, options, function (error, output, report) {
    console.log(output);
});
```

> _I have already tried setting high ulimit -n 5120, but his didn't help either._

Is there a way to pass an argument to uncss or phantomjs to process only one link at a time and when done only then move on to 2nd link?

or use promises to chain multiple uncss functions?

I have over 1000+ links that I would like to scan, I don't see this happening unless phantomjs can be sent some argument from uncss to limit number of concurrent urls to process.

Thanks

**Comment by mikelambert** [comment]
Try using uncss from head and see if that fixes your issue? I just committed a change that should serialize all phantomjs usage

**Comment by saalmaan** [comment]
fantastic, let me try that.

**Comment by RyanZim** [comment]
@mikelambert Perhaps https://github.com/giakki/uncss/blob/master/src/phantom.js#L17 has something to do with this. I haven't tested anything, just throwing out ideas. @saalmaan mentioned:

> Phantomjs instance took a lot more RAM then previous version up to 5GB

Perhaps we need to use a new PhantomJS instance when proccessing large numbers of URLs.

**Comment by mikelambert** [comment]
- The select issues should be related to https://github.com/giakki/uncss/issues/171 , I've never looked into that though.
- It shouldn't minify or compress, postcss should be producing identical css output. Are you sure there's nothing else in your config contributing to this?
- I assume the 5GB of ram might be due to running 100 pages through phantomjs and it doing a poor job of GC-ing between pages? (Perhaps they're kept in some sort of "back" history in the browser) Is your RAM usage proportional to the number of pages you pass in? My best guess is that it's related to https://github.com/ariya/phantomjs/issues/11390

Re Updates: Not sure what's going on, but I suspect it might possibly be related to the RAM issue you are referring to. And if that's the case, we might need to re-construct phantom instances (a la Ryan's suggestion) to keep it down, or experiment with something like https://github.com/ariya/phantomjs/commit/5768b705a0 API to try to clear caches and see if that helps.

**Comment by RyanZim** [comment]
Seems to me that PhantomJS should handle some of this, not us. Of course, workarounds are a necessary evil in the FOSS world.

I did a little poking around in the PhantomJS issue tracker and found this issue: https://github.com/ariya/phantomjs/issues/13581. One of the collaborators said this:

> I'm sorry we took so long to get back to you about this.

> With regret, I don't know if we're going to be able to do anything about this. It's probably some kind of memory leak or memory corruption in the interaction between Qt and Webkit, and we don't have the manpower to debug it. (I have debugged things like this before - it can take days of investigation.)

> Personally, I would recommend quitting and restarting the PhantomJS process after every page load. I use PhantomJS as a scraper that way and it works well.

> If _you_ have time to debug this and work up a patch, though, we would be glad to take it.


That was Oct 14, 2015; and there has been no maintainer action on the issue since then. :disappointed: 

It seems to me that restarting PhantomJS for each page would have a pretty big performance hit. 

Perhaps we should add an option to restart PhantomJS every _n_ pages.

CC: @mikelambert @XhmikosR

**Comment by RyanZim** [comment]
@saalmaan Have you figured out a workaround?

**Comment by saalmaan** [comment]
@RyanZim @mikelambert No, I tried this on another machine (with same settings) and got the same results, 100+ urls are fine, but the moment you want to scan 200 or 500 links everything stops working. RAM usage doesn't seems proportional to number of links. UnCSS is great tool but with this limitation (may be phantom is to be blamed) it can't be used for production large scale websites. Our eCommerce wesbite has over million pages (~30k unique, built over 10 years). Scanning for unused CSS for this many pages manually is impossible. I am more than happy to offer any help towards testing and debugging.

**Comment by RyanZim** [comment]
I don't have time to work on this right now.

A **very** ugly workaround is to split the array into ~100 item chunks and run each one seperately (output to different files), then concat the resulting files and run that file through cssnano (or another minifier) to merge the duplicate rules.

@mikelambert Thoughts on adding

> an option to restart PhantomJS every _n_ pages.

???

**Comment by RyanZim** [comment]
> I am more than happy to offer any help towards testing and debugging.

@saalmaan One thing that could be done is to verify that this is a PhantomJS problem.

Build a test script using https://github.com/peerigon/phridge. Only call `phridge.spawn()` once. Attempt to load 200+ pages and see if your script crashes.

**Comment by mikelambert** [comment]
I just pushed some changes to my personal Uncss repo that limit the PhantomJS reuse to 10 pages. @saalmaan , please give it a shot and let me know if it fixes your issues?
https://github.com/mikelambert/uncss

**Comment by RyanZim** [comment]
@saalmaan have you got a chance to try that yet?

**Comment by saalmaan** [comment]
I just tried 500+ urls and it didn't work :( i saw lots of `select: Invalid argument` and then one undefined at the end. No output. see attached screen shot
![image](https://cloud.githubusercontent.com/assets/4620730/17447168/e6716038-5b44-11e6-83c4-360296ad634b.png)

@RyanZim @mikelambert have you 500+ urls on your side?

**Comment by RyanZim** [comment]
@saalmaan A list of 500+ urls would be nice.

@mikelambert Perhaps this will shed some lignt: http://stackoverflow.com/questions/32430426/phantomjs-2-0-0-select-invalid-argument-error. Take note of the comment below the accepted answer.

**Comment by saalmaan** [comment]
here is the list of urls, I am debugging it now with node-debug

**Comment by saalmaan** [comment]
There are two instances of phantom and here is memory usage
![image](https://cloud.githubusercontent.com/assets/4620730/17451667/66a84922-5b60-11e6-91c0-629e0ceedb9a.png)

**Comment by saalmaan** [comment]
I just did a fresh install of uncss and replaced @mikelambert version with default uncss. now I don't see select: Invalid argument error, instead 

<img width="587" alt="screen shot 2016-08-06 at 05 25 42" src="https://cloud.githubusercontent.com/assets/4620730/17454667/5beadaf0-5b96-11e6-9095-a3bea9621752.png">
<img width="1022" alt="screen shot 2016-08-06 at 05 26 01" src="https://cloud.githubusercontent.com/assets/4620730/17454668/5bf1c234-5b96-11e6-9fe1-47ecac839ab8.png">

**SyntaxError: Parse error** and **INVALID_STATE_ERR: DOM Exception 11:** are coming from page, these error goes away if I try a different URL, which means these error are coming from html/js on the page, so can we ignore them?

latest phantom crash dump
[C29C7B40-9F06-4201-B76C-142053B660A0.dmp.zip](https://github.com/giakki/uncss/files/404876/C29C7B40-9F06-4201-B76C-142053B660A0.dmp.zip)

**Comment by RyanZim** [comment]
@saalmaan Do you have inline JS on any of your pages? If so, that is a known issue, see #232. In most cases, these errors will not affect your css _unless you are dynamically adding classes/identifiers onload in your inline JS_.

**Comment by mikelambert** [comment]
FWIW, I am not yet convinced that there is an issue with inline JS. (No one's presented a small use case. And in fact, it could even be possible that external relative-to-page external JS is failing to load silently, and that the erors only become visible when it's forced to run/load as part of running inline.)

However, I do fully agree that the JS errors are harmless unless the error-ing code would otherwise be adding CSS classes to elements.

@saalmaan , I am confused. Let me know if this is correct:
- On July 3rd, you tried master and had errors with memory usage
- On July 29th-ish, I wrote up a proposed fix, which you tried and found caused "select: Invalid argument"
- On August 5th, you did a fresh install of uncss (ie, from master), and saw only HTML/JS errors and not the memory usage errors that you saw on July 3rd

Except that July 3rd and August 5th should essentially be the same code except for a few minor testing/reporting-related changes. So I'm unclear why you are seeing different results...am I missing anything here?

**Comment by saalmaan** [comment]
Agreed js errors are not effecting unCSS results.
@mikelambert on July 3rd I did see phantomJS consuming high memory. Which is shown in Aug 6th comment and screen grab. I also remember at the time I did try different version of phantom to see if that helped. on 6th of Aug I decided to remove all phantom and uncss version and do a clean/fresh installs and interestingly I didnt see any `select: Invalid argument` error (even when I tried 500+ urls). I saw parsing error. I then updated unCSS with your branch and repeated the test and say same parsing error and phantom crashing after a while.

I tried node-debug on your branch, and put a breakpoint in phantom.js init function.

```
function init(reuseLimit) {
    if (phantom) {
        phantomUsed++;
        if (reuseLimit && phantomUsed >= reuseLimit) {
            phantom = null;
        }
    }
    // If we still have a phantom instance, our init() is a no-op
    if (phantom) {
        return null;
    }
    // Reset our usage counter
    phantomUsed = 0;
    // Convert to bluebird promise
    return new promise(function (resolve) {
        resolve(phridge.spawn({
            ignoreSslErrors: 'yes',
            sslProtocol: 'any'
        }));
    }).then(function (ph) {
        /* Phridge outputs everything to stdout by default */
        ph.childProcess.cleanStdout.unpipe();
        ph.childProcess.cleanStdout.pipe(process.stderr);
        phantom = ph;
    }).disposer(phridge.disposeAll);
}
```

is this function supposed to be called for each link? if its called once then how does reuseLimit enforced?

similarly uncss.js async.queue function doesn't seems to be called for each link

```
var serializedQueue = async.queue(function (opts, callback) {
    if (opts.usePostCssInternal) {
        return promise
            .using(phantom.init(opts.reuseLimit), function () {
                return getHTML(opts.html, opts)
                    .spread(processAsPostCss);
            })
            .asCallback(callback);
    }
    return promise
        .using(phantom.init(opts.reuseLimit), function () {
            return getHTML(opts.html, opts)
                .spread(getStylesheets)
                .spread(getCSS)
                .spread(processWithTextApi);
        })
        .asCallback(callback, { spread: true });
}, 1);
```

am I missing anything? also I am right in thinking that only 10 (defaultReuseLimit) links will be added to queue and when they are drained only then next 10 will be processed? this doesn't seem to be happening at the moment.

I then tried debugging by adding `"--remote-debugger-port": 8888` in phantom.js on line 38 and captured these screenshots

> help on debugging phridge and Phantom with remote debugging
> https://www.npmjs.com/package/phridge
> http://phantomjs.org/troubleshooting.html

see the screenshot below
![image](https://cloud.githubusercontent.com/assets/4620730/17466441/fc9d297c-5d07-11e6-8b2f-3b400cc0debe.png)

All these parse errors are from individual page 
see this screenshot from node-debug
![image](https://cloud.githubusercontent.com/assets/4620730/17466472/bdf1e180-5d08-11e6-9b0b-a4e229fd8c11.png)

and
![image](https://cloud.githubusercontent.com/assets/4620730/17466467/a352fa6c-5d08-11e6-9e42-09533fcc76db.png)

So you can clearly see in last screenshot that there were 70+ links, so how is this all working, 

Can you please try to run uncss on your side with the links I provided against the css file and see if you get the same results and will help you to debug it. or you can tell me next possible debugging ideas.

**Comment by saalmaan** [comment]
@mikelambert any thoughts?

**Comment by mikelambert** [comment]
Oh wow, sorry....I guess I didn't have any further thoughts at the time. :/ Didn't really have any time to contribute to this.

However, another contributor just replaced PhantomJS with JSDom, and I've released this now as 0.15.0. It's possible this will significantly help your issues, if you're still using Uncss in any capacity...

**Comment by RyanZim** [comment]
@mikelambert Can we close this?
